Site recovery support

martenbohlin opened the pull request:
This pull request adds 6 new resources needed to support site recovery (Azure to Azure).

This fixes issue #3480

Let me know if I can improve the code in any way.

martenbohlin commented:
force-pushed to fix the lint errors that I had missed.

katbyte commented on azurerm/helpers/azure/resourceid.go:
```diff
@@ -54,7 +54,8 @@ func ParseAzureResourceID(id string) (*ResourceID, error) {
 		// Catch the subscriptionID before it can be overwritten by another "subscriptions"
 		// value in the ID which is the case for the Service Bus subscription resource
-		if key == "subscriptions" && subscriptionID == "" {
+		// Also handle different casing of Subscription that happens when talking to recovery services API
```
Given this is a unique bug with the recovery services API we should handle it in the resources with a comment alluding to why, and a link to this open SDK issue: https://github.com/Azure/azure-sdk-for-go/issues/2824

katbyte commented on azurerm/resource_arm_recovery_services_network_mapping_test.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+resource "azurerm_virtual_network" "test1" {
+  name = "network1-%d"
```
Could we format the TF here and align the assingments?

katbyte commented on azurerm/resource_arm_recovery_services_protection_container.go:
```diff
@@ -0,0 +1,148 @@
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+			"name": {
```
Could we put this property up top as the first?

martenbohlin replied:
Fixed

katbyte commented on azurerm/resource_arm_recovery_services_fabric.go:
```diff
@@ -0,0 +1,145 @@
+		existing, err := client.Get(ctx, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing recovery services fabric: %+v", err)
```
For this (and all other error messages) could e be a little more clear about what is erroring:
```suggestion
				return fmt.Errorf("Error checking for presence of existing recovery services fabric %s (vault %s): %+v", name, vaultName, err)
```

katbyte commented on azurerm/resource_arm_recovery_services_protection_container.go:
```diff
@@ -0,0 +1,148 @@
+
+	future, err := client.Create(ctx, fabricName, name, parameters)
+	if err != nil {
+		return fmt.Errorf("Error creating recovery services protection container1: %+v", err)
```
As with all errors could we add a bit more detail:
```suggestion
		return fmt.Errorf("Error creating recovery services protection container %s (fabric %s): %+v", name, fabricName, err)
```

katbyte commented on azurerm/resource_arm_recovery_services_protection_container_mapping.go:
```diff
@@ -0,0 +1,171 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```
COuld we validate this is an ID?

katbyte commented on azurerm/resource_arm_recovery_services_protection_container_mapping.go:
```diff
@@ -0,0 +1,171 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```
COuld we validate this is an ID?

martenbohlin commented on azurerm/resource_arm_recovery_services_fabric.go:
```diff
@@ -0,0 +1,145 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```
Extracted this into azure.ValidateRecoveryServicesVaultName in new file azurerm/helpers/azure/recovery_services_vault.go. Also updated other resources that had duplicated this code.

martenbohlin commented on azurerm/resource_arm_recovery_services_network_mapping.go:
```diff
@@ -0,0 +1,194 @@
+				Required:         true,
+				ForceNew:         true,
+				ValidateFunc:     validate.NoEmptyStrings,
+				DiffSuppressFunc: suppress.CaseDifference,
```
Yes, the API returns another casing of the  "name" part of the ids, Thats also why I had to update line 58 of azurerm/helpers/azure/resourceid.go

katbyte replied:
Now that we are fixing this at the d.Set(id) level can we remove this?

martenbohlin replied:
Unfortunately no. The Set(id) fix handles the casing of the static word Subscription. But if I use uppercase letters in the name of my network it will come back in all lowercase.

We could lowercase the entire id before saving it, but that does not feel right.

martenbohlin commented:
Thanks @katbyte for the quick feedback. I have fixed all your comments about the code, and I will start to look at the documentation now.

It would be great if you could double check commit 63dcc2e20e324f056dfc79ecbdb5a1c6b08d0d3f where I upgrade the azure-sdk-for-go dependency, I am not sure if I updated the files under vendor correctly.

martenbohlin commented:
All requested changes has now been made.

katbyte commented on azurerm/resource_arm_recovery_services_fabric.go:
```diff
@@ -0,0 +1,147 @@
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	if props := resp.Properties; props != nil {
+		d.Set("location", props.FriendlyName) // Crazy? yes. But the location comes back in the friendly name
```
It might make sense to open an azure sdk bug for this one? 🤔

martenbohlin replied:
I found the correct way to get the location now that I have worked a bit more with the API.

katbyte commented on azurerm/resource_arm_recovery_services_network_mapping.go:
```diff
@@ -0,0 +1,194 @@
+				Required:         true,
+				ForceNew:         true,
+				ValidateFunc:     validate.NoEmptyStrings,
+				DiffSuppressFunc: suppress.CaseDifference,
```
as above

katbyte commented on azurerm/resource_arm_recovery_services_replicated_vm.go:
```diff
@@ -0,0 +1,341 @@
+	return &schema.Resource{
+		Create: resourceArmRecoveryReplicatedItemCreate,
+		Read:   resourceArmRecoveryReplicatedItemRead,
+		Update: nil,
```
This can just be removes
```suggestion
```

katbyte commented on go.mod:
```diff
@@ -1,7 +1,7 @@
 
 require (
-	github.com/Azure/azure-sdk-for-go v31.0.0+incompatible
+	github.com/Azure/azure-sdk-for-go v31.1.0+incompatible
```
If we need to update the SDK that should be done in a separate PR, i can open that and get it merged for you once 1.33 goes out 🙂

martenbohlin replied:
Thanks, that sounds great!

martenbohlin commented on azurerm/resource_arm_recovery_services_network_mapping.go:
```diff
@@ -0,0 +1,194 @@
+	return &schema.Resource{
+		Create: resourceArmRecoveryNetworkMappingCreate,
+		Read:   resourceArmRecoveryNetworkMappingRead,
+		Update: nil,
```
Fixed

martenbohlin commented:
About the test failures, I can not reproduce that. I ran `make testacc TESTARGS='-run=TestAccAzureRMRecoveryReplicationPolicy'` and get a success. And I have also tried to run the other tests that I have added.

katbyte commented on azurerm/resource_arm_recovery_services_protection_container.go:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("location", resp.Location)
```
You are setting location here and it doesn't exist in the schema

martenbohlin replied:
Aha, now I get it. Fixed.

katbyte commented on azurerm/resource_arm_recovery_services_replication_policy.go:
```diff
@@ -0,0 +1,202 @@
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("recovery_vault_name", vaultName)
+	d.Set("location", resp.Location)
```
And here

sean-nixon commented:
Super appreciative of the work on this and excited to try this out. Sorry to chime in out of nowhere but are these typos intentional? `targert_disk_type` and `targert_replica_disk_type`. I noticed them in the documentation updates as well as in the source code from the replicated VM resource. Just wanted to call that out before this gets released

martenbohlin commented:
@sean-nixon thanks for pointing that out. I have created pull request #4123 to fix the spelling mistakes.
@katbyte Will you merge that as well?
